distributed/scheduler.py

```diff
@@ -7520,9 +7520,46 @@ def decide_worker(
     objective,
     total_nthreads: Py_ssize_t,
 ) -> WorkerState:
-    """
+    r"""
     Decide which worker should take task *ts*.
 
+    There are two modes: root(ish) tasks, and normal tasks.
+
+    Root(ish) tasks
+    ~~~~~~~~~~~~~~~
+
+    Root(ish) have no (or very very few) dependencies and fan out widely:
+    they belong to TaskGroups that contain more tasks than there are workers.
+    We want neighboring root tasks to run on the same worker, since there's a
+    good chance those neighbors will be combined in a downstream operation:
+
+          i       j
+         / \     / \
+        e   f   g   h
+        |   |   |   |
+        a   b   c   d
+        \   \  /   /
+             X
+
+    In the above case, we want ``a`` and ``b`` to run on the same worker,
+    and ``c`` and ``d`` to run on the same worker, reducing future
+    data transfer. We can also ignore the location of ``X``, because
+    as a common dependency, it will eventually get transferred everywhere.
```

Comment:
<3 the ascii art

Comment/question: Do we want to explain all of this here?  Historically I haven't put the logic behind heuristics in the code.  This is a subjective opinion, and far from universal, but I find that heavily commented/documented logic makes it harder to understand the code at a glance.  I *really* like that the current decide_worker implementation fits in a terminal window.  I think that single-line comments are cool, but that long multi-line comments would better be written as documentation.  

Thoughts?  If you are not in disagreement then I would encourage us to write up a small docpage or maybe a blogpost and then link to that external resource from the code.

Reply:
I was also planning on updating https://distributed.dask.org/en/latest/scheduling-policies.html#choosing-workers, probably with this same ascii art. So just linking to that page in the docstring seems appropriate.